src/core/thread/mle.cpp

```diff
@@ -1647,11 +1647,23 @@ otError Mle::SendChildIdRequest(void)
     Message *message = NULL;
     Ip6::Address destination;
 
-    if (mRole == OT_DEVICE_ROLE_CHILD &&
-        mParent.GetExtAddress() == mParentCandidate.GetExtAddress())
+    if (mParent.GetExtAddress() == mParentCandidate.GetExtAddress())
     {
-        otLogInfoMle(GetInstance(), "Already attached to candidate parent");
-        ExitNow(error = OT_ERROR_ALREADY);
+        if (mRole == OT_DEVICE_ROLE_CHILD)
+        {
+            otLogInfoMle(GetInstance(), "Already attached to candidate parent");
+            ExitNow(error = OT_ERROR_ALREADY);
+        }
+        else
+        {
+            // Invalidate stale parent state.
+            //
+            // Parent state is not normally invalidated after becoming a Router/Leader (see #1875).  When trying to
+            // attach to a better partition, invalidating old parent state (especially when in kStateRestored) ensures
+            // that GetNeighbor() returns mParentCandidate when processing the Child ID Response.
+            mParent.SetState(Neighbor::kStateInvalid);
+            otPlatSettingsDelete(&GetInstance(), Settings::kKeyParentInfo, -1);
```

Comment:
Is there any reason/advantage to remove the old parent from the settings here? Checking the existing code, we don't do this from `BecomeDetached()`. I think the idea was that the settings entry would be overwritten once we get attached to new parent.

Reply:
Overwriting the new parent after attaching is too late in this case.  As noted in the comment, `GetNeighbor()` will return `mParent` rather than `mParentCandidate` and causes the attach to fail.

Reply:
It is not overwritten, but just written into a new logic block. And the parent information which is last written would be retrieved in `otPlatSettingsGet()`. 

The advantage may be not so obvious, but here marking the outdated information with deleted flag in flash would help to improve the efficiency of 
1) getting parent info from flash (false for this `if` to avoid more flash op) https://github.com/openthread/openthread/blob/master/examples/platforms/utils/settings.cpp#L363)
2) finding the valid information when swapping is required (false for this `if` to avoid extra loop and flash op)
 (https://github.com/openthread/openthread/blob/master/examples/platforms/utils/settings.cpp#L175)
 
It do have some subtle advantages when the device may experience multiple attach processes after long-term run.

Thoughts?

Reply:
Thanks. Sounds good to me.